### ISSUE TITLE
CLI: 8.3.0 + iVY = ERROR in Error: Cannot assign to a reference or variable!

### ISSUE THREAD

**Issue by GlauberF** [angular/angular#32695]
# 🐞 Bug report

```
- [ ] new
- [ X ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Description
I updated my CLI and enabled IVY, but when I'm building the error mentioned below.
I checked several things I found on the internet, example of the links below:

https://stackoverflow.com/questions/48021581/error-cannot-assign-to-a-reference-or-variable-angular-4

https://github.com/TrilonIO/aspnetcore-angular-universal/issues/293

And the problem, which is not pointing the location of the error, if there is error.
Because it started this only after updating the cli and activating ivy

## 🔥 Exception or Error
```
ERROR in Error: Cannot assign to a reference or variable!
    at _AstToIrVisitor.visitPropertyWrite (/var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:8412:31)
    at PropertyWrite.visit (/var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:7256:28)
    at convertActionBinding (/var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:8021:49)
    at prepareEventListenerParameters (/var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16230:27)
    at /var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:17314:24
    at /var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:17095:71
    at /var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16405:87
    at Array.map (<anonymous>)
    at TemplateDefinitionBuilder.buildTemplateFunction (/var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16405:60)
    at /var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16963:60
    at /var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16383:81
    at Array.forEach (<anonymous>)
    at TemplateDefinitionBuilder.buildTemplateFunction (/var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16383:37)
    at /var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16963:60
    at /var/www/html/pwa-vimbo/node_modules/@angular/compiler/bundles/compiler.umd.js:16383:81
    at Array.forEach (<anonymous>)
```

## 🌍 Your Environment
```javascript
Angular CLI: 8.3.0
Node: 10.15.3
OS: linux x64
Angular: 9.0.0-next.3
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, service-worker

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.803.0
@angular-devkit/build-angular      0.803.0
@angular-devkit/build-optimizer    0.803.0
@angular-devkit/build-webpack      0.803.0
@angular-devkit/core               8.3.0
@angular-devkit/schematics         8.3.0
@angular/cdk                       8.1.3
@angular/cli                       8.3.0
@angular/flex-layout               8.0.0-beta.26
@angular/material                  8.1.3
@angular/material-moment-adapter   8.1.3
@ngtools/webpack                   8.3.0
@schematics/angular                8.3.0
@schematics/update                 0.803.0
rxjs                               6.5.2
typescript                         3.5.3
webpack                            4.39.2
```

**Comment by mickz18** [comment]
Got the same error check all ref; ngFor as and let on template but didn't found. Ivy compiler should name the variable.

**Comment by cmajsmith** [comment]
@mickz18 would you mind to share the way you've managed to debug?

**Comment by mickz18** [comment]
@cmajsmith I have open ` ./node_modules/@angular/compiler/bundles/compiler.umd.js` and add `console.log(ast.name)`; 
on line 8405 (method `_AstToIrVisitor.prototype.visitPropertyWrite` just before ` throw new Error('Cannot assign to a reference or variable!'); `)

**Comment by khanghyxda** [comment]
Got the same error.
ERROR in Error: Cannot assign to a reference or variable!
    at _AstToIrVisitor.visitPropertyWrite (E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:8406:31)
    at PropertyWrite.visit (E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:7250:28)
    at convertActionBinding (E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:8015:49)
    at prepareEventListenerParameters (E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:16211:27)
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:17295:24
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:17076:71
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:16386:87
    at Array.map (<anonymous>)
    at TemplateDefinitionBuilder.buildTemplateFunction (E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:16386:60)
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:16944:60
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:16364:81
    at Array.forEach (<anonymous>)
    at TemplateDefinitionBuilder.buildTemplateFunction (E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:16364:37)
    at Object.compileComponentFromMetadata (E:\FS1\NetCore\SSite\node_modules\@angular\compiler\bundles\compiler.umd.js:17979:58)
    at ComponentDecoratorHandler.compile (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\annotations\src\component.js:496:34)
    at IvyCompilation.compileIvyFieldFor (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\compilation.js:387:57)
    at IvyVisitor.visitClassDeclaration (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\transform.js:51:40)
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\util\src\visitor.js:77:93
    at IvyVisitor.Visitor._visitListEntryNode (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\util\src\visitor.js:51:26)
    at IvyVisitor.Visitor._visit (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\util\src\visitor.js:77:36)
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\util\src\visitor.js:75:76
    at visitNodes (E:\FS1\NetCore\SSite\node_modules\typescript\lib\typescript.js:65855:48)
    at visitLexicalEnvironment (E:\FS1\NetCore\SSite\node_modules\typescript\lib\typescript.js:65888:22)
    at Object.visitEachChild (E:\FS1\NetCore\SSite\node_modules\typescript\lib\typescript.js:66194:54)
    at IvyVisitor.Visitor._visit (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\util\src\visitor.js:75:23)
    at Object.visit (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\util\src\visitor.js:25:24)
    at transformIvySourceFile (E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\transform.js:176:28)
    at E:\FS1\NetCore\SSite\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\transform.js:30:24

**Comment by alxhub** [comment]
We really need a better error case in this situation.

**Comment by Splaktar** [comment]
If a reproduction is still needed, https://github.com/angular/angular/issues/32761 (which was closed as a duplicate of this one) includes a full reproduction repo.

**Comment by Splaktar** [comment]
PR https://github.com/angular/angular/pull/32760 attempts to improve the error message, but it's still a work in progress. I don't have a solid suggestion about how this error message should be improved, but I agree that it needs improvement.

**Comment by fergalmoran** [comment]
I'm getting the same, I tried @mickz18 console.log trick - but it's not helping me

```PropertyWrite {  
  span: ParseSpan { start: 0, end: 12 },
  receiver: ImplicitReceiver { span: ParseSpan { start: 0, end: 0 } },
  name: 'entry',      
  value: PropertyRead { 
    span: ParseSpan { start: 6, end: 12 }, 
    receiver: ImplicitReceiver { span: [ParseSpan] },
    name: '$event'  
  } 
}
```
I renamed every $event var in my templates but the error is still the same (I'm guessing maybe a 3rd party lib?) - could really do with some better errors here.

**Comment by ShinyGeodavid** [comment]
For anyone who couldn't get the `console.log(ast.name)`  in the compiler.umd.js file to work like me. You can just change line 8406 where it throws the error from `throw new Error('Cannot assign to a reference or variable!');` to `throw new Error('Cannot assign to a reference or variable! ' + ast.name);`. Then your reference will be easy to find in the error name. My error came from something like this.
```typescript
<div *ngFor="let item of items">
	<mat-form-field>
		<input matInput [(ngModel)]="item">
	</mat-form-field>
</div>
```
Changing to html to this fixed it...
```typescript
<div *ngFor="let item of items; let i = index">
	<mat-form-field>
		<input matInput [(ngModel)]="items[i]">
	</mat-form-field>
</div>
```

**Comment by JoostK** [comment]
The error has been changed into an error diagnostic that shows exactly where the invalid write occurs in #34339. As such, I'm closing this issue as resolved.
